kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -388,50 +387,66 @@ private Map<String, DomainMetadata> loadDomainMetadataMap(Engine engine) {
   }
 
   /**
-   * Calculates the latest snapshot hint before or at the current snapshot version, returns the
+   * Calculates the latest snapshot hint before or at the current snapshot version, cache the
    * CRCInfo if checksum file at the current version is read
+   *
+   * <p>Strategy for finding and using CRC (checksum) files:
+   *
+   * <ol>
+   *   <li>During initial log directory listing, the SnapshotManager captures all files including
+   *       commits, checkpoints, and checksums.
+   *   <li>The most recent checksum file discovered is stored in the LogSegment as
+   *       "lastSeenChecksum", regardless of its version, however, logSegment will make sure
+   *       last_check_point_version <= crc_version <= logSegment_version
+   *   <li>When loading table state information like Protocol, Metadata, or Domain Metadata:
+   *       <ol>
+   *         <li>First check if there's a lastSeenChecksum in the LogSegment and if its version is
+   *             appropriate (>= lower bound, <= requested version)
+   *         <li>If a suitable checksum exists, generate version hint directly from it for
+   *             subsequence call use that
+   *         <li>If no suitable checksum exists, subsequence call fall back to reading from log
+   *             files
+   *       </ol>
+   * </ol>
    */
-  private Tuple2<Optional<SnapshotHint>, Optional<CRCInfo>>
-      maybeGetNewerSnapshotHintAndCurrentCrcInfo(
-          Engine engine,
-          LogSegment logSegment,
-          Optional<SnapshotHint> snapshotHint,
-          long snapshotVersion) {
-
-    // Snapshot hint's version is current.
+  private Optional<SnapshotHint> maybeGetNewerSnapshotHintAndUpdateCrcInfoCache(
+      Engine engine,
+      LogSegment logSegment,
+      Optional<SnapshotHint> snapshotHint,
+      long snapshotVersion) {
+
+    // Snapshot hint's version is current so we could use it in loading P&M.
+    // No need to read crc.
     if (snapshotHint.isPresent() && snapshotHint.get().getVersion() == snapshotVersion) {
-      return new Tuple2<>(snapshotHint, Optional.empty());
+      return snapshotHint;
```

Comment:
Can we move the `if (snapshotHint.isPresent() && snapshotHint.get().getVersion() > snapshotVersion) {` to the constructor? and just consider this part of the constructor input validation?